OutOfSchool/OutOfSchool.WebApi/Services/SubordinationStructure/InstitutionHierarchyService.cs

```diff
@@ -0,0 +1,146 @@
+        /// <param name="localizer">Localizer.</param>
+        /// <param name="mapper">Mapper.</param>
+        public InstitutionHierarchyService(
+            ISensitiveEntityRepository<InstitutionHierarchy> repository,
+            IWorkshopRepository repositoryWorkshop,
+            IProviderRepository repositoryProvider,
+            ILogger<InstitutionHierarchyService> logger,
+            IStringLocalizer<SharedResource> localizer,
+            IMapper mapper)
+        {
+            this.repository = repository ?? throw new ArgumentNullException(nameof(repository));
+            this.repositoryWorkshop = repositoryWorkshop;
+            this.repositoryProvider = repositoryProvider;
+            this.localizer = localizer;
+            this.logger = logger;
+            this.mapper = mapper;
+        }
+
+        /// <inheritdoc/>
+        public async Task<InstitutionHierarchyDto> Create(InstitutionHierarchyDto dto)
+        {
+            logger.LogInformation("InstitutionHierarchy creating was started.");
+
+            var institutionHierarchy = mapper.Map<InstitutionHierarchy>(dto);
+
+            var newInstitutionHierarchy = await repository.Create(institutionHierarchy).ConfigureAwait(false);
+
+            logger.LogInformation($"InstitutionHierarchy with Id = {newInstitutionHierarchy?.Id} created successfully.");
+
+            return mapper.Map<InstitutionHierarchyDto>(newInstitutionHierarchy);
+        }
+
+        /// <inheritdoc/>
+        public async Task<Result<InstitutionHierarchyDto>> Delete(Guid id)
+        {
+            logger.LogInformation($"Deleting InstitutionHierarchy with Id = {id} started.");
+
+            var entity = new InstitutionHierarchy() { Id = id };
+
+            try
+            {
+                await repository.Delete(entity).ConfigureAwait(false);
+
+                logger.LogInformation($"InstitutionHierarchy with Id = {id} succesfully deleted.");
+
+                return Result<InstitutionHierarchyDto>.Success(mapper.Map<InstitutionHierarchyDto>(entity));
+            }
+            catch (DbUpdateConcurrencyException)
+            {
+                logger.LogError($"Deleting failed. InstitutionHierarchy with Id = {id} doesn't exist in the system.");
+                throw;
+            }
+        }
+
+        /// <inheritdoc/>
+        public async Task<IEnumerable<InstitutionHierarchyDto>> GetAll()
+        {
+            logger.LogInformation("Getting all InstitutionHierarchies started.");
+
+            var institutionHierarchies = await repository.GetAll().ConfigureAwait(false);
+
+            logger.LogInformation(!institutionHierarchies.Any()
+                ? "InstitutionHierarchy table is empty."
+                : $"All {institutionHierarchies.Count()} records were successfully received from the InstitutionHierarchy table.");
+
+            return institutionHierarchies.Select(entity => mapper.Map<InstitutionHierarchyDto>(entity)).ToList();
```

Comment:
Please be consistent - either change method result type to `Task<List<....` or don't do premature materialization in return using `.ToList()`. This might affect performance a bit.

Reply:
Similar point for `GetChildren`

Reply:
Done